File: RetrofitReservation/app/src/main/java/ma/ensa/retrofitreservation/service/ApiInterface.java
package ma.ensa.retrofitreservation.service;

import java.util.List;

import ma.ensa.retrofitreservation.models.Chambre;
import ma.ensa.retrofitreservation.models.Client;
import ma.ensa.retrofitreservation.models.Reservation;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface ApiInterface {

    // Endpoints pour Chambre
    @GET("api/chambres")
    Call<List<Chambre>> getAllChambres();

    @POST("api/chambres")
    Call<Chambre> createChambre(@Body Chambre chambre);

    @GET("api/chambres/{id}")
    Call<Chambre> getChambreById(@Path("id") Long id);

    @PUT("api/chambres/{id}")
    Call<Chambre> updateChambre(@Path("id") Long id, @Body Chambre chambre);

    @DELETE("api/chambres/{id}")
    Call<Void> deleteChambre(@Path("id") Long id);

    // Endpoints pour Client
    @GET("api/clients")
    Call<List<Client>> getAllClients();

    @GET("api/clients/names")
    Call<List<String>> getClientsNames();

    @POST("api/clients")
    Call<Client> createClient(@Body Client client);

    @GET("api/clients/{id}")
    Call<Client> getClientById(@Path("id") Long id);

    @PUT("api/clients/{id}")
    Call<Client> updateClient(@Path("id") Long id, @Body Client client);

    @DELETE("api/clients/{id}")
    Call<Void> deleteClient(@Path("id") Long id);

    // Ajouter une méthode dans votre API pour récupérer un client par son nom et prénom
    @GET("clients/find")
    Call<Client> getClientByNameAndSurname(@Query("name") String name, @Query("surname") String surname);

    // Récupérer toutes les réservations
    @GET("api/reservations")
    Call<List<Reservation>> getAllReservations();

    // Créer une nouvelle réservation
    @POST("api/reservations")
    Call<Reservation> createReservation(@Body Reservation reservation);

    // Récupérer une réservation par ID
    @GET("api/reservations/{id}")
    Call<Reservation> getReservationById(@Path("id") Long id);

    // Mettre à jour une réservation par ID
    @PUT("api/reservations/{id}")
    Call<Reservation> updateReservation(@Path("id") Long id, @Body Reservation reservation);

    // Supprimer une réservation par ID
    @DELETE("api/reservations/{id}")
    Call<Void> deleteReservation(@Path("id") Long id);
}

File: ReservationRetrofit_KOTLIN/app/src/main/java/ma/ensaj/reservationretrofit_kotlin/service/ApiInterface.kt
package ma.ensaj.reservationretrofit_kotlin.service

import ma.ensaj.reservationretrofit_kotlin.models.Chambre
import ma.ensaj.reservationretrofit_kotlin.models.Client
import ma.ensaj.reservationretrofit_kotlin.models.Reservation
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query
interface ApiInterface {

    // Endpoints pour Chambre
    @GET("api/chambres")
    fun getAllChambres(): Call<List<Chambre>>

    @POST("api/chambres")
    fun createChambre(@Body chambre: Chambre): Call<Chambre>

    @GET("api/chambres/{id}")
    fun getChambreById(@Path("id") id: Long): Call<Chambre>

    @PUT("api/chambres/{id}")
    fun updateChambre(@Path("id") id: Long, @Body chambre: Chambre): Call<Chambre>

    @DELETE("api/chambres/{id}")
    fun deleteChambre(@Path("id") id: Long): Call<Void>

    // Endpoints pour Client
    @GET("api/clients")
    fun getAllClients(): Call<List<Client>>

    @GET("api/clients/names")
    fun getClientsNames(): Call<List<String>>

    @POST("api/clients")
    fun createClient(@Body client: Client): Call<Client>

    @GET("api/clients/{id}")
    fun getClientById(@Path("id") id: Long): Call<Client>

    @PUT("api/clients/{id}")
    fun updateClient(@Path("id") id: Long, @Body client: Client): Call<Client>

    @DELETE("api/clients/{id}")
    fun deleteClient(@Path("id") id: Long): Call<Void>

    // Ajouter une méthode dans votre API pour récupérer un client par son nom et prénom
    @GET("clients/find")
    fun getClientByNameAndSurname(@Query("name") name: String, @Query("surname") surname: String): Call<Client>

    // Récupérer toutes les réservations
    @GET("api/reservations")
    fun getAllReservations(): Call<List<Reservation>>

    // Créer une nouvelle réservation
    @POST("api/reservations")
    fun createReservation(@Body reservation: Reservation): Call<Reservation>

    // Récupérer une réservation par ID
    @GET("api/reservations/{id}")
    fun getReservationById(@Path("id") id: Long): Call<Reservation>

    // Mettre à jour une réservation par ID
    @PUT("api/reservations/{id}")
    fun updateReservation(@Path("id") id: Long, @Body reservation: Reservation): Call<Reservation>

    // Supprimer une réservation par ID
    @DELETE("api/reservations/{id}")
    fun deleteReservation(@Path("id") id: Long): Call<Void>

}
